src/sec.c

```diff
@@ -23,10 +23,15 @@ sec_bin_init(sec_bin_t *bin) {
 bool
 sec_init(tsdn_t *tsdn, sec_t *sec, base_t *base, pai_t *fallback,
     const sec_opts_t *opts) {
-	size_t max_alloc = opts->max_alloc & PAGE_MASK;
-	pszind_t npsizes = sz_psz2ind(max_alloc);
-	if (sz_pind2sz(npsizes) > opts->max_alloc) {
-		npsizes--;
+	assert(opts->max_alloc > 0);
+
+	pszind_t npsizes = sz_psz2ind(opts->max_alloc);
+	/*
+	 * If max_alloc falls exactly on the size class boundary, we need one
+	 * more bin.
```

Comment:
I'm a bit lost here. Isn't the size class inclusive?

Reply:
For some reason, I was doing it under the assumption the all extents we would cache had sizes exactly matching the size class boundaries (obviously incorrect). IDK why I thought so.

Let's reiterate - when we do `sz_psz2ind`, it returns the index of the smallest page size class, which covers `max_alloc`. The number of bins should be equal to this number + 1, because the indices are zero-based.

Before the change, we were subtracting one if the size class boundary is larger than `opts->max_alloc`. I don't quite understand why, though. I think I need David's help with that, because I can't figure it out. To me, it seems that we should always add 1.

Some examples (if we always do sz_psz2ind of truncated opts->max_alloc and add one):
* opts->max_alloc = 4096 => max_alloc = 4096, sz_psz2ind = 0, npsizes = 1
* opts->max_alloc = 4097 => max_alloc = 4096, sz_psz2ind = 0, npsizes = 1
* opts->max_alloc = 36900 => max_alloc = 36864, sz_psz2ind = 8, npsizes = 9. Here, I should clarify: the 7th bin is 32768, the 8th bin is 40960. Even though `sz_pind2sz(npsizes) > opts->max_alloc` is true, I don't see why we need to do `npsizes--`, because extent with size `36864` (exactly between two size classes) should also be cached, and it falls into the 8th bin.

Reply:
> all extents we would cache had sizes exactly matching the size class boundaries (obviously incorrect)

Is it? I think this was right, and that the original subtraction of 1 is just a bug that never got noticed because the other bug covered it up.

Are we all agreed that the actual logic should be:
1. Round opts->max_alloc down to a page size to obtain the effective max size.
2. Make that the maximum pszind_t we can access (so that the number of bins is that + 1)
?

I think just reimplementing that from scratch will be easier than trying to make sense of whatever nonsense I initially wrote.

Reply:
> all extents we would cache had sizes exactly matching the size class boundaries (obviously incorrect)
Yes, we can cache extents with size falling in between the boundaries. For instance, this happens when you do a large allocation and add sz_large_pad. This only happens if you tweak the default `hpa_sec_max_alloc`.

> Are we all agreed that the actual logic should be
Yes, I think that's the correct approach here.